### PR TITLE
correct ExactSizeIterator examples

### PR DESCRIPTION
**Pull request by kineticz**

### CODE REVIEW

File: src/libcore/iter/traits.rs [hunk 1]
```diff
@@ -473,15 +473,15 @@ impl<'a, I: DoubleEndedIterator + ?Sized> DoubleEndedIterator for &'a mut I {
 /// impl ExactSizeIterator for Counter {
 ///     // We already have the number of iterations, so we can use it directly.
 ///     fn len(&self) -> usize {
-///         self.count
+///         5 - self.count
```
Comment (apasel422) [inline src/libcore/iter/traits.rs]: Won't this overflow when `self.count == 6`? Seems like `self.count` should only be incremented in the `next` function in the `Some` branch.

Comment (kindlychung) [inline src/libcore/iter/traits.rs]: Yeah, that and `6 - self.count()`, like this:

```
/// # struct Counter {
/// #     count: usize,
/// # }
/// # impl Counter {
/// #     fn new() -> Counter {
/// #         Counter { count: 0 }
/// #     }
/// # }
/// # impl Iterator for Counter {
/// #     type Item = usize;
/// #     fn next(&mut self) -> Option<usize> {
/// #         if self.count < 6 {
/// #             self.count += 1;
/// #             Some(self.count)
/// #         } else {
/// #             None
/// #         }
/// #     }
/// # }
/// impl ExactSizeIterator for Counter {
///     // We already have the number of iterations, so we can use it directly.
///     fn len(&self) -> usize {
///         6 - self.count
///     }
/// }
///
/// // And now we can use it!
///
/// let counter = Counter::new();
///
/// assert_eq!(6, counter.len());
/// assert_eq!(Some(1), counter.next());
/// assert_eq!(5, counter.len());
/// assert_eq!(Some(2), counter.next());
/// assert_eq!(4, counter.len());
/// assert_eq!(Some(3), counter.next());
/// assert_eq!(3, counter.len());
/// assert_eq!(Some(4), counter.next());
/// assert_eq!(2, counter.len());
/// assert_eq!(Some(5), counter.next());
/// assert_eq!(1, counter.len());
/// assert_eq!(Some(6), counter.next());
/// assert_eq!(0, counter.len());
/// assert_eq!(None, counter.next());
```

**Comment by kindlychung** [comment]
A newbie question: How do I run the assertions the the example locally?

`cargo test` doesn't take care of these.

**Comment by bluss** [comment]
Good question, I'm not sure which target it would be. Can't we shorten the example code, we don't need all the assertions?

**Comment by bluss** [comment]
It's not super easy to dig out of the log, but the test does indeed fail:

```
failures:

---- iter::traits::ExactSizeIterator_1 stdout ----
	<anon>:35:21: 35:28 error: cannot borrow immutable local variable `counter` as mutable
<anon>:35 assert_eq!(Some(1), counter.next());
                              ^~~~~~~
<anon>:35:21: 35:28 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
<anon>:37:21: 37:28 error: cannot borrow immutable local variable `counter` as mutable
<anon>:37 assert_eq!(Some(2), counter.next());
                              ^~~~~~~
<anon>:37:21: 37:28 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
<anon>:39:21: 39:28 error: cannot borrow immutable local variable `counter` as mutable
<anon>:39 assert_eq!(Some(3), counter.next());
                              ^~~~~~~
<anon>:39:21: 39:28 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
<anon>:41:21: 41:28 error: cannot borrow immutable local variable `counter` as mutable
<anon>:41 assert_eq!(Some(4), counter.next());
                              ^~~~~~~
<anon>:41:21: 41:28 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
<anon>:43:21: 43:28 error: cannot borrow immutable local variable `counter` as mutable
<anon>:43 assert_eq!(Some(5), counter.next());
                              ^~~~~~~
<anon>:43:21: 43:28 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
<anon>:45:21: 45:28 error: cannot borrow immutable local variable `counter` as mutable
<anon>:45 assert_eq!(Some(6), counter.next());
                              ^~~~~~~
<anon>:45:21: 45:28 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
<anon>:47:18: 47:25 error: cannot borrow immutable local variable `counter` as mutable
<anon>:47 assert_eq!(None, counter.next());
                           ^~~~~~~
<anon>:47:18: 47:25 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:32:5: 32:12 help: to make the local variable mutable, use `mut` as shown:
<anon>:   let mut counter = Counter::new();
error: aborting due to previous error(s)
thread 'iter::traits::ExactSizeIterator_1' panicked at 'Box<Any>', src/librustc/session/mod.rs:167
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    iter::traits::ExactSizeIterator_1


```

File: src/libcore/iter/traits.rs [hunk 2]
```diff
@@ -462,8 +462,8 @@ impl<'a, I: DoubleEndedIterator + ?Sized> DoubleEndedIterator for &'a mut I {
 /// # impl Iterator for Counter {
 /// #     type Item = usize;
 /// #     fn next(&mut self) -> Option<usize> {
-/// #         self.count += 1;
 /// #         if self.count < 6 {
+/// #             self.count += 1;
```
Comment (bluss) [inline src/libcore/iter/traits.rs]: Reducing the number in the conditional to `5` would keep the length the same. Keeping 5 seems nice. Needs adjustment to 5 in size_hint.

File: src/libcore/iter/traits.rs [hunk 3]
```diff
@@ -473,15 +473,28 @@ impl<'a, I: DoubleEndedIterator + ?Sized> DoubleEndedIterator for &'a mut I {
...
 ///     // We already have the number of iterations, so we can use it directly.
 ///     fn len(&self) -> usize {
-///         self.count
+///         6 - self.count
 ///     }
 /// }
 ///
 /// // And now we can use it!
 ///
 /// let counter = Counter::new();
```
Comment (bluss) [inline src/libcore/iter/traits.rs]: needs to be `mut`. I would use just two assertions here, we don't need them all (it's supposed to be example, instructive code, not an exhaustive test).

**Comment by bluss** [comment]
I think the test target for running doc tests may be either of these: `make check-stage1-doc-crate-core` or `make check-stage2-doc-crate-core`.

You can also use `TESTNAME=ExactSizeIterator` to have it run just the tests matching a string.

**Comment by steveklabnik** [comment]
@kindlychung ping! Did @bluss answer your question? Any interest in keeping up this PR?

**Comment by steveklabnik** [comment]
Closing due to a lack of activity; just let me know if you find the time to fix it up :smile:

**Pull request closed without merging**